
---
title: "AnalysisOutputs"
author: "FlicAnderson"
date: "03/08/2020"
output: html_document
params:
  output_dir: "./"  # Output directory
  orf_fasta_file: FALSE  # FASTA file with nt seq
  orf_gff_file: NA  # riboviz generated GFF2/GFF3 annotation file
  num_processes: 1  # Number of cores for parallelization
  min_read_length: 10  # Minimum read length in H5 output
  max_read_length: 50  # Maximum read length in H5 output
  buffer: 250  # Length of flanking region around the CDS
  primary_id: "gene_id"  # Primary gene IDs to access the data (YAL001C, YAL003W, etc.)
  dataset: "vignette"  # Name of the dataset
  rpf: TRUE  # Is the dataset an RPF or mRNA dataset?
  features_file: NA  # features file, columns are gene features and rows are genes
  do_pos_sp_nt_freq: TRUE  # do calculate the position-specific nucleotide frequency
  t_rna_file: NA  # tRNA estimates in .tsv file
  codon_positions_file: NA  # Codon positions in each gene in .Rdata file
  count_threshold: 64  # threshold for count of reads per gene to be included in plot
  length_threshold: 500  # threshold for length of genes to be included in metagene plots
  output_prefix: ""  # Prefix for output files
  hd_file: "output.h5"  # Location of H5 output file
  nnt_buffer: 25  # n nucleotides of UTR buffer to include in metagene plots
  nnt_gene: 50  # n nucleotides of gene to include in metagene plots
  asite_disp_length_file: NA  # asite displacement file table with one displacement per read length
---

The outputs can be changed by editing the params called in the following terminal command, 
... which should match the params listed in the .yaml section at the top of the .Rmd

`Rscript -e "rmarkdown::render(\"rmarkdown/AnalysisOutputs.Rmd\", params=list(orf_fasta_file=\"/vignette/input/yeast_YAL_CDS_w_250utrs.fa\", orf_gff_file=\"/vignette/input/yeast_YAL_CDS_w_250utrs.gff3\", min_read_length=10, max_read_length=50, dataset=\"vignette\", hd_file=\"/vignette/output/WT3AT/WT3AT.h5\"))"`

From the python call to generate_stats_figs.R (for sample WT3AT from vignette):
> Rscript --vanilla /home/flic/riboviz/rscripts/generate_stats_figs.R --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=vignette/output/WT3AT/WT3AT.h5 --orf-fasta-file=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --output-dir=vignette/output/WT3AT --do-pos-sp-nt-freq=True --t-rna-file=data/yeast_tRNAs.tsv --codon-positions-file=data/yeast_codon_pos_i200.RData --features-file=data/yeast_features.tsv --orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=data/yeast_standard_asite_disp_length.txt --count-threshold=64

To run the thing from inside R, use `rmarkdown::render("rmarkdown/AnalysisOutputs.Rmd", params=list(orf_fasta_file="/vignette/input/yeast_YAL_CDS_w_250utrs.fa", orf_gff_file="/vignette/input/yeast_YAL_CDS_w_250utrs.gff3", min_read_length=10, max_read_length=50, dataset="vignette", hd_file="/vignette/output/WT3AT/WT3AT.h5"))`


File path variables need to be wrapped in here() to give full path, or it doesn't work (e.g. orf_fasta_file, orf_gff_file, hd_file).

```{r, setup, include=FALSE}
# set this option in the first code chunk in the document
knitr::opts_chunk$set(echo = params$my_name)
```

```{r setup2, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)


# load getopt to allow use of get_Rscript_filename for provenance-gathering
# and sourcing read_count_functions.R
# load here for the same reason
suppressMessages(library(getopt, quietly=T))
suppressMessages(library(here))
suppressMessages(library(Rsamtools))
suppressMessages(library(rtracklayer))
suppressMessages(library(rhdf5))
suppressMessages(library(parallel))
suppressMessages(library(optparse))
suppressMessages(library(RcppRoll))
suppressMessages(library(ggplot2))
suppressMessages(library(tidyr))
suppressMessages(library(dplyr))
suppressMessages(library(magrittr))
suppressMessages(library(purrr))
suppressMessages(library(here))
```

```{r source_script, include=FALSE}

source(here::here("rscripts", "provenance.R"), local = knitr::knit_global())
source(here::here("rscripts", "read_count_functions.R"), local = knitr::knit_global())
source(here::here("rscripts", "stats_figs_block_functions.R"), local = knitr::knit_global())

# or sys.source("your-script.R", envir = knitr::knit_global())
```

```{r close_connections_1}
# prepare files, opens hdf5 file connection
#hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
rhdf5::h5closeAll()
```

```{r provenance}
# handle interactive vs standard running
if(interactive()){
  this_script <- "AnalysisOutputs.Rmd"
  path_to_this_script <- here("rmarkdown", this_script)
} else {
  this_script <- getopt::get_Rscript_filename()
  path_to_this_script <- this_script
}

# print provenance
print_provenance(path_to_this_script)
```


```{r read_ins}

#print(params$my_name)

# print(my_name)
# # label: read_ins
# # Quitting from lines 89-120 (AnalysisOutputs.Rmd) 
# # Error in print(my_name) : object 'my_name' not found

# # head(h5ls(params$hd_file, recursive = FALSE)) 
# # # doesn't work with: rmarkdown::render("rscripts/AnalysisOutputs.Rmd", params=list(hd_file="vignette/output/WT3AT/WT3AT.h5"))
# 
# head(h5ls(here(params$hd_file), recursive = FALSE))
# # does work with: rmarkdown::render("rscripts/AnalysisOutputs.Rmd", params=list(hd_file="vignette/output/WT3AT/WT3AT.h5"))
# # therefore need here() round stuff!
#   # can I do that at the params bit?  SEEMS NOT :P  Suppose this is because I haven't library(here)'d yet.

# read in positions of all exons/genes in GFF format and subset CDS locations
gene_names <- rhdf5::h5ls(here(params$hd_file), recursive = 1)$name
head(gene_names)

# # read in coding sequences
coding_seqs <- Biostrings::readDNAStringSet(here(params$orf_fasta_file))
str(coding_seqs)
# 
# # range of read lengths between parameters set in config file
read_range <- params$min_read_length:params$max_read_length
read_range
# 
# # read in positions of all exons/genes in GFF format and convert to tibble data frame
gff_df <- readGFFAsDf(here(params$orf_gff_file))
str(gff_df)

# set ggplot2 theme for plots drawn after this; use dark on light theme
ggplot2::theme_set(theme_bw())

hd_file <- here(params$hd_file)

```

```{r check_params}

print(params)
str(params)

attach(params)

#three_nucleotide_periodicity_data <- CalculateThreeNucleotidePeriodicity(gene_names = gene_names, dataset = dataset, hd_file = hd_file, gff_df = gff_df)

# three_nucleotide_periodicity_data <- CalculateThreeNucleotidePeriodicity(gene_names = gene_names, dataset = params$dataset, hd_file = here(params$hd_file), gff_df = gff_df)
# 
# str(three_nucleotide_periodicity_data)


```


```{r 3nt}
#ThreeNucleotidePeriodicity <- function(gene_names, dataset, hd_file, gff_df) {

  # check for 3nt periodicity
  print("Starting: Check for 3nt periodicity globally")

  # CalculateThreeNucleotidePeriodicity():
  three_nucleotide_periodicity_data <- CalculateThreeNucleotidePeriodicity(gene_names = gene_names, dataset = dataset, hd_file = hd_file, gff_df = gff_df)

  # PlotThreeNucleotidePeriodicity()
  PlotThreeNucleotidePeriodicity(three_nucleotide_periodicity_data)

  # NOTE: repeated from inside CalculateThreeNucleotidePeriodicity() as preferred not to return multiple objects in list (hassle :S)
  gene_poslen_counts_5start_df <- AllGenes5StartPositionLengthCountsTibble(gene_names = gene_names, dataset= dataset, hd_file = hd_file, gff_df = gff_df)

  # run PlotStartCodonRiboGrid()
  start_codon_ribogrid_plot <- PlotStartCodonRiboGrid(gene_poslen_counts_5start_df)
  # creates plot object
  start_codon_ribogrid_plot

  # run SaveStartCodonRiboGrid():
  #SaveStartCodonRiboGrid(start_codon_ribogrid_plot)

  # run PlotStartCodonRiboGridBar():
  start_codon_ribogrid_bar_plot <- PlotStartCodonRiboGridBar(gene_poslen_counts_5start_df)
  # creates plot object
  start_codon_ribogrid_bar_plot
  
  # run SaveStartCodonRiboGridBar():
  #SaveStartCodonRiboGridBar(start_codon_ribogrid_bar_plot)

  # run SavePlotThreeNucleotidePeriodicity():
  #SavePlotThreeNucleotidePeriodicity(three_nucleotide_periodicity_plot)

  # run WriteThreeNucleotidePeriodicity():
  #WriteThreeNucleotidePeriodicity(three_nucleotide_periodicity_data)
# Quitting from lines 160-204 (AnalysisOutputs.Rmd) 
# Error in file(data_file, open = "w") : cannot open the connection

  print("Completed: Check for 3nt periodicity globally")

#} # end ThreeNucleotidePeriodicity() function definition

# run ThreeNucleotidePeriodicity():
#ThreeNucleotidePeriodicity(gene_names, dataset, hd_file, gff_df)

```


```{r close_connections_2}
# prepare files, opens hdf5 file connection
#hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
rhdf5::h5closeAll()
detach(params)
rm(list=ls())
```
