---
title: "RiboViz Processing"
output: 
  flexdashboard::flex_dashboard:
    theme: yeti
    orientation: rows
    vertical_layout: fill
runtime: shiny
---

<!-- Using a JS hack to store the currently selected tab into a reactive value -->
<script>
$("body").on("shown.bs.tab", "a[data-toggle='tab']", function(e) {
   Shiny.setInputValue("active_tab", $(e.target).parent().index());
})
</script>

```{r packages}
library(flexdashboard) # Also imports shiny
library(tidyverse)
library(plotly)
library(ggfittext) # Also imports ggplot2
library(patchwork) # To display multiple plots in one tabset tab
```

### Read in files

This section is dedicated to reading in all of the files.
```{r}
# read counts file
read_counts <- read_tsv("../../output_nf/read_counts.tsv", skip=5)
```

Read length dists, eventually collpased to one df
```{r}
rl.locs <- dir("../../output_nf", recursive = TRUE, pattern = "read_lengths.tsv", full.names = TRUE) 


# Set the names of this vector to be just the sample names
# fix this, i only want the folder/sample name not tsv
names(rl.locs) <- sapply(rl.locs, function(x){
  strsplit(x, "/") %>% 
    unlist() %>% 
    .[[4]]
  })
```



```{r loading}
# Loading read_counts.tsv
counts_df <- read_tsv("../../output_nf/read_counts.tsv", skip=5) %>% # First 5 rows are metadata
  slice(-3) %>% # After column names, next 3 rows are input, cudadapt, and Unassigned
  filter(SampleName != "NA") # Removing NA values

# Loading read_lengths.tsv
# Find all matching files in a dir
rl.locs <- dir("/home/tilton/code/PET_RiboViz/replicate_data", recursive = TRUE, pattern = "read_lengths.tsv", full.names = TRUE) 
# Set the names of this vector to be just the sample names
names(rl.locs) <- sapply(rl.locs, function(x){
  strsplit(x, "/") %>% 
    unlist() %>% 
    .[[7]]
})
# Combine into properly labeled data
lengths_df <- bind_rows(lapply(rl.locs, read_tsv, skip = 4), .id="sample")

# Loading 3nt_periodicity.tsv
# Same method as read_lengths.tsv
per.locs <- dir("/home/tilton/code/PET_RiboViz/replicate_data", recursive = TRUE, pattern = "3nt_periodicity.tsv", full.names = TRUE)
names(per.locs) <- sapply(per.locs, function(x){
  strsplit(x, "/") %>% 
    unlist() %>% 
    .[[7]]
})
periodicity_df <- bind_rows(lapply(per.locs, read_tsv, skip = 4), .id="sample")

# Loading 3ntframe_bygene.tsv
# Same method as read_lengths.tsv
byg.locs <- dir("/home/tilton/code/PET_RiboViz/replicate_data", recursive = TRUE, pattern = "3ntframe_bygene.tsv", full.names = TRUE)
names(byg.locs) <- sapply(byg.locs, function(x){
  strsplit(x, "/") %>% 
    unlist() %>%  
    .[[7]]
})
# For bar chart
frame_ribo <- bind_rows(lapply(byg.locs, read_tsv, skip = 4), .id="sample") 
#%>%
 # select(sample, Ct_fr0, Ct_fr1, Ct_fr2)
frame_bar <- frame_ribo %>%
  select(sample, Ct_fr0, Ct_fr1, Ct_fr2) %>%
  group_by(sample) %>%
  summarise(Frame_0 = sum(Ct_fr0),
            Frame_1 = sum(Ct_fr1),
            Frame_2 = sum(Ct_fr2))
  #aggregate(cbind(tmp$Ct_fr0, tmp$Ct_fr1, tmp$Ct_fr2), by=list(sample=tmp$sample), FUN=sum) %>%
 # rename(Ct_fr0 = V1, Ct_fr1 = V2, Ct_fr2 = V3)
#

frame_box <- frame_ribo %>%
  #rowwise() %>%
  mutate(Frame_0 = Ct_fr0/(Ct_fr0 + Ct_fr1 + Ct_fr2), 
         Frame_1 = Ct_fr1/(Ct_fr0 + Ct_fr1 + Ct_fr2),
         Frame_2 = Ct_fr2/(Ct_fr0 + Ct_fr1 + Ct_fr2)) %>%
  select(sample, gene, Frame_0, Frame_1, Frame_2) %>%
  pivot_longer(names_to = "Frame", cols = 3:5)
```

```{r setup}
# Colors for plots
a_color <- c("#FF1F58",
             "#00CD6C",
             "#009ADE",
             "#AF58BA",
             "#F28522")

# Preparing name swap for counts_df
a <- c("Total reads", "Mapped to rRNAs", "Total unaligned with rRNA", "Mapped to ORF", "Deduplicated reads")
rcdf_descs <- c("Demultiplexed reads", 
                "rRNA or other contaminating reads removed by alignment to rRNA index files", 
                "Unaligned reads removed by alignment of remaining reads to ORFs index files", 
                "Reads aligned to ORFs index files", 
                "Deduplicated reads")
label_swap_df <- data_frame(Description=rcdf_descs, NewLabels=a)

# Drop-down options for "Single Group" selection size
group_names <- c("WT", "ncs2", "uba4")

# Drop-down options for "Single Replicate" selection size
replicate_names <- unique(counts_df[,1])[[1]]
```

```{r theme}
plot_theme <- theme(panel.background = element_blank(),
                    panel.grid = element_blank(),
                    text = element_text(size = 13),
                    axis.line = element_line(color = "black"))

plot_color = "steelblue"
```

```{r plots}
# Dodged bar plot for reads that got through each step
counts_dodged <- reactive({
  counts_df_formatted() %>% 
      ggplot(., mapping=aes(x=reorder(NewLabels, NumReads), y=NumReads, text=paste("Number of Reads:", NumReads)))+
      geom_bar(stat="identity", fill=plot_color)+
      geom_bar_text()+
      labs(title = paste("Demultiplex Counts per Step:", samp()))+
      xlab(label=NULL)+
      theme(legend.position="none")+
      coord_flip()+
      plot_theme
})

# Histogram for count of reads at a given length
lengths_plot <- reactive({
  lengths_df %>% 
    filter(grepl(samp(), sample)) %>% 
    ggplot(., aes(x = Length, y = Counts)) +
    geom_bar(stat = "identity", fill=plot_color) +
    labs(y="Count" ,title = paste("Read Length Counts:", samp()))+
    plot_theme
})

# 3nt_periodicity.tsv
# Line plot
# Position on x
# Counts on y
# Facet grid by end (5' or 3')
periodicity_plot <- reactive({
  periodicity_df %>% 
    filter(grepl(samp(), sample)) %>%
    group_by(Pos, End) %>%
    summarize(Counts=sum(Counts)) %>%
    ggplot(., aes(x = Pos, y = Counts)) + 
    geom_line(color=plot_color) + 
    facet_grid(rows = vars(End), scales="free") + 
    labs(title = paste("Periodicity:", samp()))+
    plot_theme
})

# 3ntframe_bygene.tsv
# Sum all rows for cols 2-4
# Bar chart for frame 0, 1, 2
frames_plot <- reactive({
  frame_bar %>% 
    filter(grepl(samp(), sample)) %>% 
    select(Frame_0, Frame_1, Frame_2) %>%
    colSums() %>%
    data.frame(Frame=names(.), Count=., row.names=NULL) %>%
    ggplot(., aes(x = Frame, y = Count)) + 
    geom_bar(stat="identity", fill=plot_color) + 
    labs(title = paste("Frames 0-2:", samp()))+
    plot_theme
})

# 3ntframe_bygene.tsv
# Box plot
frame_boxplot <- reactive({
  frame_box %>%
    filter(grepl(samp(), sample)) %>% 
    ggplot(., aes(x=Frame, y=value)) +
    labs(y="Proportion of Reads")+
    geom_boxplot(fill=plot_color)+
    plot_theme
})

# Two plots ripped from read_count_functions.R
# Currently missing frame_grid
ribogrid_plot <- reactive({
  filter(grepl(samp(), sample)) %>% 
  ggplot(., aes(x = Pos, y = ReadLen, fill = Counts)) +
    geom_tile() +
    scale_fill_gradient("count", low = "white", high = "darkblue") +
    theme(panel.grid = element_blank()) +
    labs(x = "position of read 5' end", y = "read length")
})
ribogrid_barplot <- reactive({
  filter(grepl(samp(), sample)) %>% 
  ggplot(
    data = filter(., ReadLen %in% small_read_range),
    aes(x = Pos, y = Counts)
  ) +
    geom_col() +
    facet_grid(ReadLen ~ ., scales = "free_y") +
    scale_y_continuous(expand = c(0, 0)) +
    labs(x = "position of read 5' end", y = "count")
})


# Put TPM plot code here
```

```{r reactive}
# Getting the desired sample name
samp <- reactive({
  if(is.na(input$samp)) {
    "*"
  } else {
    paste(input$samp)  
  }
})

# Creating a dataframe for visualization of counts_df 
# This is a nightmare, I would like to make it neater
counts_df_formatted <- reactive({
  counts_df %>%
    select(SampleName, NumReads, Description) %>%
    filter(grepl(samp(), SampleName)) %>% 
    filter(as.character(Description) != "Reads after trimming of 5' mismatches and removal of those with more than 2 mismatches" 
        & as.character(Description) != "Reads with rRNA and other contaminating reads removed by alignment to rRNA index files") %>% 
    left_join(., label_swap_df, by="Description") %>% # Combine filters here
    group_by(NewLabels) %>% 
    summarise(NumReads = sum(NumReads)) %>%
    mutate(dummy="Demultiplex")
})
```

## Input {.sidebar}

```{r input}
# Display UI depending on selected output tab
output$input_ui <- renderUI({
   # See JS code at top to see where active_tab comes from
   sel <- input$active_tab
   
   # UI for "All Samples"
   if (is.null(sel) || sel == 0) {
     radioButtons("samp", "All Samples", c("*")) # I'm not sure how to set a reactive value directly
   
   # UI for "Single Group"
   } else if (sel == 1) {
      radioButtons("samp", "Group Name:", group_names, group_names[1])
    
   # UI for "Single Replicate"
   } else {
      radioButtons("samp", "Replicate Name:", replicate_names)        
   }
})

# Show the prepared UI
uiOutput("input_ui")
```

## Data {.tabset}

### All Samples
```{r output_allsamples}
renderPlot({
  plot((counts_dodged() + lengths_plot()) / (periodicity_plot() + frames_plot()) / (frame_boxplot()))
  # / (ribogrid_plot() + ribogrid_barplot())
})
```

### Single Group
```{r output_singlegroup}
renderPlot({
  plot((counts_dodged() + lengths_plot()) / (periodicity_plot() + frames_plot()) / frame_boxplot())
  # / (ribogrid_plot() + ribogrid_barplot())
})
```

### Single Replicate
```{r output_singlereplicate}
renderPlot({
  plot((counts_dodged() + lengths_plot()) / (periodicity_plot() + frames_plot())  / frame_boxplot())
  # / (ribogrid_plot() + ribogrid_barplot())
})
```